submariner-operator/issue-149-node-placement.md

```diff
@@ -0,0 +1,96 @@
+# Node Placement
+
+[Add support for nodeSelector and toleration scheduling for submariner pods](https://github.com/submariner-io/enhancements/issues/149)
+
+## Summary
+
+Submariner deploys many different agents as part of its core deployment. Gateway, Globalnet Agent and
+Metrics Proxy are deployed on designated gateway nodes, while Route Aents are deployed on all nodes. There is no way
+for users to control or limit where rest of the agent pods are deployed. Some customers want to run these agents on
+specific nodes for billing or networking reasons. Users can't add tolerations directly to these agents coz they'll be
+overwritten by the operator. So, users require an easily configurable way to specify such configurations.
+
+## Proposal
+
+This enhancement proposes adding a new field to `Submariner` and `ServiceDiscovery` for users to configure node
+scheduling. This configuration will then be applied to pods of following submariner agents:
+
+* submariner-operator
+* submariner-lighthouse-agent
+* submariner-lighthouse-coredns
+
+Following agents will be excluded because these are run on Gateway nodes. Users already control which nodes get labeled
+as gateways and ideally we don't want to run anything else on these nodes.
+
+* submariner-gateway
+* submariner-globalnet
+* submariner-metrics-proxy
+
+## Design Details
+
+Add a new `NodePlacement` field to `Submariner` and `ServiceDiscovery` specs.
+
+```Go
+// NodePlacement describes node scheduling configuration for the pods.
+type NodePlacement struct {
+    // NodeSelector defines which Nodes the Pods are scheduled on. The default is an empty list.
+    // +optional
+    NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+    // Tolerations is attached by pods to tolerate any taint that matches
+    // the triple <key,value,effect> using the matching operator <operator>.
+    // The default is an empty list.
+    // +optional
+    Tolerations []v1.Toleration `json:"tolerations,omitempty"`
+}
+```
+
+```Go
+type SubmarinerSpec struct {
+	...
+	...
+	// NodePlacement enables explicit control over the scheduling of the submariner pods.
+	// +optional
+	NodePlacement NodePlacement `json:"nodePlacement,omitempty"`
+}
+```
+
+```Go
+type SubmarinerDiscoverySpec struct {
+	...
+	...
+	// NodePlacement enables explicit control over the scheduling of the submariner pods.
+	// +optional
+	NodePlacement NodePlacement `json:"nodePlacement,omitempty"`
+}
+```
+
+Add option to `subctl join` to specify node selectors and tolerations.
+
+```bash
+subctl join --node-selectors LABEL_KEY_1=LABEL_VALUE_1,...,LABEL_KEY_N=LABEL_VALUE_N
+subctl join --node-tolerations KEY_1=VAL_1:TAINT_EFFECT_1,...,KEY_N=VAL_N:TAINT_EFFECT_N
```

Comment:
This is a very simplistic way of representing these, which can also include matchers and other info (eg toleration seconds).

I wonder if we shouldn't just let users give a yaml (string or file) for this?

Reply:
I used `kubectl` cli for reference. Yaml/json string or file does sound like a better more flexible option, while this seems easier one.

Reply:
Do you have a reference to where in `kubectl` this is available? I couldn't find it

Reply:
```
Usage:
  kubectl taint NODE NAME KEY_1=VAL_1:TAINT_EFFECT_1 ... KEY_N=VAL_N:TAINT_EFFECT_N [options]

Usage:
  kubectl label [--overwrite] (-f FILENAME | TYPE NAME) KEY_1=VAL_1 ... KEY_N=VAL_N [--resource-version=version]
[options]
```

Reply:
Ok but that is for the taints, I haven't seen any example where tolerations are given on the command line, only as YAML examples (which I guess you're supposed to edit/apply/patch)

Reply:
@vthapar :point_up: